lib/xcore.js

```diff
@@ -31,14 +31,20 @@ function XClient(displayNum, screenNum, options)
 
     this.displayNum = displayNum;
     this.screenNum = screenNum;
-    this.authHost = os.hostname();
 }
 util.inherits(XClient, EventEmitter);
 
 XClient.prototype.init = function(stream)
 {
     this.stream = stream;
 
+    this.authHost = stream.remoteAddress;
+    this.authFamily = stream.remoteFamily;
+    if (!this.authHost || this.authHost === '127.0.0.1' || this.authHost === '::1') {
+      this.authHost = os.hostname();
+      this.authFamily = null;
+    }
```

Comment:
because key is hostname, not ip or 'localhost'. Here is the dump of my Xauthority:

```
laplace@laplace-VirtualBox:~$ hexdump -C .Xauthority
00000000  01 00 00 12 6c 61 70 6c  61 63 65 2d 56 69 72 74  |....laplace-Virt|
00000010  75 61 6c 42 6f 78 00 01  30 00 12 4d 49 54 2d 4d  |ualBox..0..MIT-M|
00000020  41 47 49 43 2d 43 4f 4f  4b 49 45 2d 31 00 10 a8  |AGIC-COOKIE-1...|
00000030  3d b3 df 0c b3 c9 54 e7  c1 c4 8c 48 63 97 fa 01  |=.....T....Hc...|
00000040  00 00 12 6c 61 70 6c 61  63 65 2d 56 69 72 74 75  |...laplace-Virtu|
00000050  61 6c 42 6f 78 00 02 31  30 00 12 4d 49 54 2d 4d  |alBox..10..MIT-M|
00000060  41 47 49 43 2d 43 4f 4f  4b 49 45 2d 31 00 10 ea  |AGIC-COOKIE-1...|
00000070  3d fd ef 6e 08 6b 9e b9  36 2b 8d 9d 81 40 95     |=..n.k..6+...@.|
0000007f

laplace@laplace-VirtualBox:~$ hostname
laplace-VirtualBox
```